// cassandra-orm/lib/settings.js
/**
 *  Copyright 2012 Rackspace
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */


/**
 * Settings which are populated during the runtime after the initialize function
 * has been called.
 */
exports.HOSTS = [];
exports.KEYSPACE = null;
exports.USE_BIG_INTS = false;
exports.TIMEOUT = 4000;
exports.QUERY_TIMEOUT = null;
exports.CONNECTION_CLASS = null;
exports.DEFAULT_READ_CONSISTENCY = 'ONE';
exports.DEFAULT_WRITE_CONSISTENCY = 'ONE';
exports.INDEX_CLEANUP_WHITELIST = [];
exports.ACCOUNTING_CF = null;
exports.TRACKED_MODELS = [];
exports.LOG_FUNC = null;
exports.LOG_REWRITER_FUNC = null;
exports.POOLED_CONNECTION_CLS = null;
exports.CQL_VERSION = null;
exports.SOFT_REMOVE_TTL = null;


/**
 * Path to the directory containing migrations.
 * @type {String}
 */
exports.MIGRATIONS_PATH = null;


/**
 * Path to the directory containing models.
 * @type {String}
 */
exports.MODELS_PATH = null;
